pl_examples/domain_templates/reinforce_learn_ppo.py

```diff
@@ -0,0 +1,477 @@
+            self.state = self.state.to(device=self.device)
+
+            with torch.no_grad():
+                pi, action, value = self(self.state)
+                log_prob = self.actor.get_log_prob(pi, action)
+
+            next_state, reward, done, _ = self.env.step(action.cpu().numpy())
+
+            self.episode_step += 1
+
+            self.batch_states.append(self.state)
+            self.batch_actions.append(action)
+            self.batch_logp.append(log_prob)
+
+            self.ep_rewards.append(reward)
+            self.ep_values.append(value.item())
+
+            self.state = torch.FloatTensor(next_state)
+
+            epoch_end = step == (self.steps_per_epoch - 1)
+            terminal = len(self.ep_rewards) == self.max_episode_len
+
+            if epoch_end or done or terminal:
+                # if trajectory ends abtruptly, boostrap value of next state
+                if (terminal or epoch_end) and not done:
+                    self.state = self.state.to(device=self.device)
+                    with torch.no_grad():
+                        _, _, value = self(self.state)
+                        last_value = value.item()
+                        steps_before_cutoff = self.episode_step
+                else:
+                    last_value = 0
+                    steps_before_cutoff = 0
+
+                # discounted cumulative reward
+                self.batch_qvals += self.discount_rewards(self.ep_rewards + [last_value], self.gamma)[:-1]
+                # advantage
+                self.batch_adv += self.calc_advantage(self.ep_rewards, self.ep_values, last_value)
+                # logs
+                self.epoch_rewards.append(sum(self.ep_rewards))
+                # reset params
+                self.ep_rewards = []
+                self.ep_values = []
+                self.episode_step = 0
+                self.state = torch.FloatTensor(self.env.reset())
+
+            if epoch_end:
+                train_data = zip(
+                    self.batch_states, self.batch_actions, self.batch_logp,
+                    self.batch_qvals, self.batch_adv)
+
+                for state, action, logp_old, qval, adv in train_data:
+                    yield state, action, logp_old, qval, adv
+
+                self.batch_states.clear()
+                self.batch_actions.clear()
+                self.batch_adv.clear()
+                self.batch_logp.clear()
+                self.batch_qvals.clear()
+
+                # logging
+                self.avg_reward = sum(self.epoch_rewards) / self.steps_per_epoch
+
+                # if epoch ended upbruptly, exlude last cut-short episode to prevent stats skewness
+                if not done:
+                    total_epoch_reward = sum(self.epoch_rewards[:-1])
+                    nb_episodes = len(self.epoch_rewards) - 1
+                else:
+                    total_epoch_reward = sum(self.epoch_rewards)
+                    nb_episodes = len(self.epoch_rewards)
```

Comment:
There is room for deduplication in this code.
like
```python
epoch_rewards = self.epoch_rewards
if not done:
    epoch_rewards = epoch_rewards[:-1]
total_epoch_reward = sum(epoch_rewards)
nb_episodes = len(epoch_rewards)
```

Reply:
Yup this is better